EIPS/eip-5805.md

```diff
@@ -0,0 +1,401 @@
+
+### Methods
+
+#### now
+
+This function returns the current timepoint. It could be `block.timestamp`, `block.number` (or any other **non-decreasing** function) depending on the mode the contract is operating on.
+
+- If operating using **block number**, then this function SHOULD be implemented.
+- If operating using **timestamp**, then this function MUST be implemented.
+- If operating using any other mode, then this function MUST be implemented.
+
+This function is thus optional, and its absence should be considered as a marker of the contract operating using block number. (This makes this EIP compatible with pre-existing voting contracts)
+
+```yaml
+- name: now
+  type: function
+  stateMutability: view
+  inputs: []
+  outputs:
+    - name: timepoint
+      type: uint256
+```
+
+#### getVotes
+
+This function returns the current voting weight of an account. This corresponds to all the voting power delegated to it at the moment this function is called.
+
+As tokens delegated to `address(0)` should not be counted/snapshoted, `getVotes(0)` SHOULD always return `0`.
+
+This function MUST be implemented
+
+```yaml
+- name: getVotes
+  type: function
+  stateMutability: view
+  inputs:
+    - name: account
+      type: address
+  outputs:
+    - name: votingWeight
+      type: uint256
+```
+
+#### getPastVotes
+
+This function returns the historical voting weight of an account. This corresponds to all the voting power delegated to it at a specific timepoint. The timepoint parameter should match the operating mode of the contract. This function SHOULD only serve past checkpoints that are immutable. Calling this function with a timepoint that is greater or equal to `now()` SHOULD revert. For any timepoint that is strictly smaller than `now()`, the value returned by `getPastVotes` should be constant.
+
+As tokens delegated to `address(0)` should not be counted/snapshoted, `getPastVotes(0,x)` SHOULD always return `0` (for all values of `x`).
+
+This function MUST be implemented
+
+```yaml
+- name: getPastVotes
+  type: function
+  stateMutability: view
+  inputs:
+    - name: account
+      type: address
+    - name: timepoint
+      type: uint256
+  outputs:
+    - name: votingWeight
+      type: uint256
+```
+
+#### delegates
+
+This function returns the address to which the voting power of an account is currently delegated.
+
+Note that if the delegate is `address(0)` then the voting power SHOULD NOT be checkpointed, and it should not be possible to vote with it.
+
+This function MUST be implemented
+
+```yaml
+- name: delegates
+  type: function
+  stateMutability: view
+  inputs:
+    - name: account
+      type: address
+  outputs:
+    - name: delegatee
+      type: address
+```
+
+#### delegate
+
+This function changes the caller's delegate, updating the vote delegation in the meantime.
+
+This function MUST be implemented
+
+```yaml
+- name: delegate
+  type: function
+  stateMutability: nonpayable
```

Comment:
I wonder if it's intentional to be `nonpayable`?

Reply:
Yes. 

Would you change that?

Reply:
There are two reasons. For what I know, solidity _add_ restriction when we mark it non-payable. It's more gas efficient if put in a payable rather than non-payable. 


See the solidity source code on how the Type address is implemented: the NonPayable checks callvalue 
https://github.com/ethereum/solidity/blob/5f8b48793cdf9a41d6f16fa601250987bee82e3e/libsolidity/codegen/ContractCompiler.cpp#L429
See conversation here

> Adding a payable keyword also lowers the number of opcodes being executed (no check for msg.value == 0), which lowers the gas consumption. Many production use-cases will include it in the final contract design for gas efficiency reasons even if it doesn't use msg.value, assuming there is no security concern with accepting it.

https://github.com/ethereum/EIPs/pull/5758#issuecomment-1287889709

Reply:
I believe case 2 will probably need additional data, or at least additional logic, and it should not use this function signature. This function allows the sender (and only the sender) to change its delegation. If you want another mechanism, that allows a third party to pay something to get the delegation, that should be a completely different function.

For someone to change its own delegation, there is no reason to pay anything. If a payment is involved, then again, its a different mechanism that should use a different function signature.

Also, EIPs are about security, not gas effectiveness. I personally think making function payable to save gas is a very bad thing that will just lead to lost funds.

Just like ERC20's "approve" is not payable, this should also not be payable.

Reply:
LGTM!. While I reserve my opinion on this particular point but respect this is author's discretion at the technical choice made in this proposal.

That said, this different understanding about ERC20's `approval` whether they mean to be `payable` or `nonpayable` or _unspecified_ seem to hard to be represented in Solidity, and hence a good point to use ABI actually